processing/src/main/java/io/druid/query/dimension/ForwardingFilteredDimensionSelector.java

```diff
@@ -0,0 +1,200 @@
+              if (forwardedValue >= 0) {
+                // Make the following check after the `forwardedValue >= 0` check, because if forwardedValue is -1 and
+                // valueId is -1, we don't want to return true from matches().
+                if (forwardedValue == valueId) {
+                  return true;
+                }
+                nullRow = false;
+              }
+            }
+            // null should match empty rows in multi-value columns
+            return nullRow && matchNull;
+          }
+        };
+      } else {
+        return BooleanValueMatcher.of(false);
+      }
+    } else {
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          final IndexedInts baseRow = selector.getRow();
+          final int baseRowSize = baseRow.size();
+          boolean nullRow = true;
+          for (int i = 0; i < baseRowSize; i++) {
+            int rowValueId = baseRow.get(i);
+            int forwardedValue = forwardMapping.get(rowValueId);
+            if (forwardedValue >= 0) {
+              if (Objects.equals(selector.lookupName(rowValueId), value)) {
+                return true;
+              }
+              nullRow = false;
+            }
+          }
+          // null should match empty rows in multi-value columns
+          return nullRow && matchNull;
+        }
+      };
+    }
+  }
+
+  @Override
+  public ValueMatcher makeValueMatcher(Predicate<String> predicate, final boolean matchNull)
+  {
+    final BitSet valueIds = DimensionSelectorUtils.makePredicateMatchingSet(this, predicate);
+    return new ValueMatcher()
+    {
+      @Override
+      public boolean matches()
+      {
+        final IndexedInts baseRow = selector.getRow();
+        final int baseRowSize = baseRow.size();
+        boolean nullRow = true;
+        for (int i = 0; i < baseRowSize; ++i) {
+          int forwardedValue = forwardMapping.get(baseRow.get(i));
+          if (forwardedValue >= 0) {
+            if (valueIds.get(forwardedValue)) {
+              return true;
+            }
+            nullRow = false;
+          }
+        }
+        // null should match empty rows in multi-value columns
+        return nullRow && matchNull;
+      }
+    };
+  }
+
+  @Override
+  public int getValueCardinality()
+  {
+    return forwardMapping.size();
+  }
+
+  @Override
+  public String lookupName(int id)
+  {
+    return selector.lookupName(reverseMapping[id]);
+  }
+
+  @Override
+  public boolean nameLookupPossibleInAdvance()
+  {
+    return true;
+  }
+
+  @Nullable
+  @Override
+  public IdLookup idLookup()
+  {
+    final IdLookup baseIdLookup = selector.idLookup();
+    if (baseIdLookup != null) {
+      return this;
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public int lookupId(String name)
+  {
+    // Don't cache selector.idLookup(), because it is likely implemented as `return this`, in this case caching only
```

Comment:
For my own learning, what's the JVM behavior that results in the caching being slower when idLookup() returns `this`?